src/analysis.rs

```diff
@@ -99,6 +99,19 @@ impl Cost {
         self.0 <= weight.saturating_mul(1000)
     }
 
+    /// Return the minimum budget required to cover the cost.
+    fn required_budget(&self) -> u32 {
+        // TODO: Wait for rust 1.73.0
+        // self.0.div_ceil(1000)
+        let base = self.0 / 1000;
+        let remainder = self.0 % 1000;
+        if remainder > 0 {
+            base.saturating_add(1)
+        } else {
+            base
+        }
+    }
```

Comment:
In f09a045c552aaddda7181dd92b1e007e06bf9089:

This would be much clearer if the whole function were just `base.saturating_add(999) / 1000`. With a comment saying that `saturating_add` is just there to eliminate the panic path and that we don't expect it to ever saturate.

Reply:
`self.0.saturating_add(999) / 1000` would be cleaner, but its semantics are different:
`4294967001.saturating_add(999) / 1000 = 4294967` while `4294967001.div_ceil(1000) = 4294968`. The correct budget is `ceil(4294967.001) = 4294968`, which is also what `required_budget` computes.

I think it is important to compute the budget correctly, even though it exceeds `CONSENSUS_MAX`. We can use `div_ceil` once we have Rust 1.73.0.

Reply:
So, the only difference is when we're nearing u32::MAX? But even your code does the wrong thing when we are literally at the maximum.

Reply:
If we're going to use `saturating_add` at all then we are accepting that it's ok to get wrong answers when we are near the limits of the type. And in this case, my code is clearer.

Reply:
We can remove `saturating_add` in my code. It's unnecessary because we divided `base` by 1000.

I think `self.0.saturating_add(999) / 1000` is a good option.

Let me just clarify that my code is equivalent to `div_ceil` even at `u32::MAX`: [Rust playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1d40eb1a7e7d30768e1ba202b924e3e0)

Reply:
I believe that it's equivalent but it has probably 100x the cognitive load for readers of the code.

I'm fine with the existing code (though I'd prefer you remove `saturating_add` which is simply confusing, if it can never saturate) if you add a comment explaining that it's equivalent to the idiomatic `(x + 999) / 1000` except that it handles 999 more values correctly.